tfhe/src/integer/gpu/server_key/radix/oprf.rs

```diff
@@ -445,174 +235,250 @@ impl CudaServerKey {
+                    LweBskGroupingFactor(0),
+                    self.message_modulus,
+                    self.carry_modulus,
+                    PBSType::Classical,
+                    message_bits_count,
+                    total_random_bits as u32,
+                    d_bsk.d_ms_noise_reduction_key.as_ref(),
                 );
             }
-            CudaBootstrappingKey::MultiBit(d_multibit_bsk) => {
-                cuda_multi_bit_programmable_bootstrap_lwe_ciphertext(
-                    &ct_seeded,
-                    &mut ct_out.as_mut().d_blocks,
-                    &d_accumulator,
-                    &d_lut_vector_indexes,
-                    &d_output_indexes,
-                    &d_input_indexes,
-                    d_multibit_bsk,
+            CudaBootstrappingKey::MultiBit(d_bsk) => {
+                grouped_oprf_async(
                     streams,
+                    result,
+                    &d_seeded_lwe_input,
+                    num_active_blocks as u32,
+                    num_blocks as u32,
+                    &d_bsk.d_vec,
+                    d_bsk.input_lwe_dimension,
+                    d_bsk.glwe_dimension,
+                    d_bsk.polynomial_size,
+                    self.key_switching_key.decomposition_level_count(),
+                    self.key_switching_key.decomposition_base_log(),
+                    d_bsk.decomp_level_count,
+                    d_bsk.decomp_base_log,
+                    d_bsk.grouping_factor,
+                    self.message_modulus,
+                    self.carry_modulus,
+                    PBSType::MultiBit,
+                    message_bits_count,
+                    total_random_bits as u32,
+                    None,
                 );
             }
         }
+    }
 
-        let plaintext_to_add = (p - 1) * delta / 2;
-        let ct_cloned = ct_out.duplicate(streams);
-        unsafe {
-            add_lwe_ciphertext_vector_plaintext_scalar_async(
+    // Getter for the GPU memory usage of OPRF.
+    //
+    pub fn get_par_generate_oblivious_pseudo_random_unsigned_integer_size_on_gpu(
+        &self,
+        streams: &CudaStreams,
+    ) -> u64 {
+        match &self.bootstrapping_key {
+            CudaBootstrappingKey::Classic(d_bsk) => get_programmable_bootstrap_size_on_gpu(
                 streams,
-                &mut ct_out.as_mut().d_blocks.0.d_vec,
-                &ct_cloned.as_ref().d_blocks.0.d_vec,
-                plaintext_to_add,
-                out_lwe_dimension,
-                num_ct_blocks as u32,
-            );
+                d_bsk.input_lwe_dimension,
+                d_bsk.glwe_dimension,
+                d_bsk.polynomial_size,
+                d_bsk.decomp_level_count,
+                1,
+                d_bsk.d_ms_noise_reduction_key.as_ref(),
+            ),
+            CudaBootstrappingKey::MultiBit(d_bsk) => {
+                get_programmable_bootstrap_multi_bit_size_on_gpu(
```

Comment:
I think this isn't up to date anymore, there should be a function that calls the scratch with allocate_gpu_memory = false in integer/gpu/mod.rs